File: apps/mobile/lib/core/networking/api_response_envelope.dart
import 'api_error_envelope.dart';
import 'api_meta.dart';

typedef ApiJsonDecoder<T> = T Function(Object? json);

class ApiResponseEnvelope<T> {
  const ApiResponseEnvelope({
    required this.success,
    required this.meta,
    this.data,
    this.error,
  });

  factory ApiResponseEnvelope.fromJson(
    Map<String, dynamic> json, {
    required ApiJsonDecoder<T> decodeData,
  }) {
    final success = json['success'] as bool? ?? false;
    final meta = ApiMeta.fromJson(json['meta'] as Map<String, dynamic>?);

    if (success) {
      return ApiResponseEnvelope<T>(
        success: true,
        data: decodeData(json['data']),
        meta: meta,
      );
    }

    return ApiResponseEnvelope<T>(
      success: false,
      error: json['error'] is Map<String, dynamic>
          ? ApiErrorEnvelope.fromJson(json['error'] as Map<String, dynamic>)
          : const ApiErrorEnvelope(
              code: 'UNKNOWN_ERROR',
              message: 'Unknown API error',
            ),
      meta: meta,
    );
  }

  final bool success;
  final T? data;
  final ApiErrorEnvelope? error;
  final ApiMeta meta;
}

File: apps/admin-dashboard/core/api/api-response-envelope.ts
import { parseApiErrorEnvelope, type ApiErrorEnvelope } from './api-error-envelope';
import { parseApiMeta, type ApiMeta } from './api-meta';

export type ApiJsonDecoder<T> = (value: unknown) => T;

export type ApiSuccessEnvelope<T> = {
  readonly success: true;
  readonly data: T;
  readonly meta: ApiMeta;
};

export type ApiFailureEnvelope = {
  readonly success: false;
  readonly error: ApiErrorEnvelope;
  readonly meta: ApiMeta;
};

export type ApiResponseEnvelope<T> =
  | ApiSuccessEnvelope<T>
  | ApiFailureEnvelope;

export function parseApiResponseEnvelope<T>(
  value: unknown,
  decodeData: ApiJsonDecoder<T>,
): ApiResponseEnvelope<T> {
  if (!isObject(value)) {
    return {
      success: false,
      error: {
        code: 'INVALID_RESPONSE',
        message: 'Backend API returned a non-object response.',
      },
      meta: {},
    };
  }

  const meta = parseApiMeta(value.meta);

  if (value.success === true) {
    return {
      success: true,
      data: decodeData(value.data),
      meta,
    };
  }

  return {
    success: false,
    error: parseApiErrorEnvelope(value.error),
    meta,
  };
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}
